Parallel tests

badpandabear opened the pull request:
This is a large set of changes to run the dev suite test setups in parallel, add some reporting options, and allow for pytest testing of the pypeit test script.

- The pypeit_test script was split with most of the code going into test_scripts/pypeit_test.py. This allowed for unit tests in test_pypeit_test.py to import the main function and to test it with mocked subprocess code.  The pytest tests take two minutes to get 96% coverage of pypeit_test.py, instead of taking hours to discover a syntax error at the very end of a test run.

- Test setups can be run in parallel from a PriorityQueue.  The test setup was encapsulated into a TestSetup class that can categorizing setups based on how fast they run. The slowest tests are run first for efficient CPU utilization. If the slower tests were run later in the test run they could leave CPUs idle once the shorter tests have completed.  The tests are roughly categorized into SHORT, MEDIUM, LONG, VERY_LONG, and EXTREMELY_LONG categories.

- Test types have been encapsulated into a class hierarchy with common code to run tests in the base PypeItTest class.

- Reporting logic is separated out into a TestReport class that reports on testing as it runs and can also generate test reports.

- There is a -r option now for writing a detailed test report to a file.  The detailed report can also be printed to stdout if the -v (verbose) option is given. 

- There is a -q option to suppress all output to stdout so that testing can be run in the background with a detailed being written to a file.

Including a detailed test report and the default stdout from a test run.

[pype_test_output.txt](https://github.com/pypeit/PypeIt-development-suite/files/5215354/pype_test_output.txt)

[two_threads_sep11.report.txt](https://github.com/pypeit/PypeIt-development-suite/files/5215343/two_threads_sep11.report.txt)

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+        missing_data = []
+        missing_pypfiles = []
+        for instr in instruments:
+            # Only do blue instruments
```
Is this just for testing?

badpandabear replied:
The --debug option predated my changes, but I assume it was intended for testing. I considered removing it but I found it useful as a smaller set of tests to run.

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+                if setup not in setups:
+                    missing_data += [ '{0}/{1}'.format(instr, setup) ]
+
+                if instr == 'shane_kast_blue' and '600' in setup:
```
Can we generalize this?  By that, I mean add to one of the classes above a set of `setup` instruments.  And we may add to that if we want to do this for others?  As opposed to this hack I put in long ago..

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+
+    # TODO: Include option that forces the tests to run
+    # pypeit_setup
+    if setup.instr == 'shane_kast_blue' and '600' in setup.name:
```
Same as above.

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+        # ----------------------------------------------------------
+        # SensFunc tests
+        # ----------------------------------------------------------
+        # UVIS algorithm without .sens file
```
This comment is out of date.  The test below is on shane_kast_blue, not vlt_xshooter

badpandabear replied:
I fixed the comment. But should there be a UVIS test?

profxj replied:
I think not.

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+        # SensFunc tests
+        # ----------------------------------------------------------
+        # UVIS algorithm without .sens file
+        if setup.instr == 'shane_kast_blue' and '600' in setup.name:
```
I think it would be better to have a list somewhere (up above) 
of instruments that we wish to test fluxing on.  And, if necessary,
also include these `std_file` names.  I'd prefer, though, to have the
name be eg. only `Feige66` and the code would search the `spec1d`
files for any with `Feige66` in it and use the first one.   In case we 
change aspects of the naming convention (possible).

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+        # ----------------------------------------------------------
+        # fluxing/coadd1d/tellfit tests
+        # ----------------------------------------------------------
+        flux_coadd1d_pairs = {'shane_kast_blue': '600_4310_d55',  # long-slit
```
I like this and wonder if it should be set at the top so we don't have
to dig into here to modify.

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+            setup.tests.append(PypeItCoadd2DTest(setup, obj='pisco'))
+
+        # MultiSlit
+        if setup.instr == 'keck_lris_blue' and setup.name == 'multi_600_4000_d560':
```
I'd also like to remove these `if` statements and just
have `coadd2d` run on a list defined up above.

profxj commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+    ql_tests = []
+
+    # MOS
+    if setup.instr == 'shane_kast_blue' and '600' in setup.name:
```
Same.  We should have a list of QL modes that this then just runs on.

profxj reviewed (changes requested):
Great stuff.  My main request is that we endeavor to replace 
the `if/then` statements that drive the flux, coadd, telluric tests in `build_test_setup()`
with lists/dicts defined near the top of `pypeit_test.py` that then
are parsed by `build_test_setup()`.  Akin to how we somewhat seemlessly
add a new instrument to the `develop` dict.

The other requests are that you provide:

1.  A brief HOWTO on how to add new instruments or new tests
2.  Guidelines on how much RAM is required to run with a given number of processors.

Thanks!

badpandabear commented on test_scripts/pypeit_test.py:
```diff
@@ -0,0 +1,1159 @@
+
+
+    def _get_test_counts(self):
+        """Helper method to create a string with the curren test counts"""
```
fixed

badpandabear commented:
With the latest set of changes I split the pypeit_test file into 

- test_main.py: The frame work for running tests
- pypeit_tests.py: The classes used to run different test scripts
- test_setups.py: The definition for what setups and instruments are part of the dev suite, and which tests to run on what setups. 

As discussed in the slack, I also added a test_priority_list which contains the test setups in the dev suite in decreasing order of duration.  It is re-written every time a full run of the dev suite finishes, if anything has changed.  This replaces the earlier TestLength logic.

profxj commented on test_scripts/test_pypeit_test.py:
```diff
@@ -25,12 +25,12 @@
 """
```
Add a short description of the main purpose of this module.

profxj commented on test_scripts/test_setups.py:
```diff
@@ -0,0 +1,218 @@
+   the development suite google drive. See https://pypeit.readthedocs.io/en/latest/development.html#development-suite
+2) Add the instrument and setup(s) to the develop_setups dict
+3) If this is a new instrument, add the instrument to the supported_instruments list.
+3) If additional tests are add 'instrument/setup' to the desired test attribute.
```
missing a word, e.g. 'desired' ?

profxj commented on test_scripts/test_setups.py:
```diff
@@ -0,0 +1,218 @@
+                 {'std_file': 'spec1d_**GD71*.fits'},
+             }
+
+_flux_setup = ['shane_kast_blue/600_4310_d55',
```
fanatastic!

profxj reviewed (approved):
approved

profxj commented:
Accidentally added two commits to this branch/PR.
Please ignore them.  They are benign.
